modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/DistributionZoneRebalanceEngine.java

```diff
@@ -0,0 +1,286 @@
+                                        replicas,
+                                        evt.entryEvent().newEntry().revision(),
+                                        metaStorageManager,
+                                        part,
+                                        tableAssignments.isEmpty() ? Collections.emptySet() : tableAssignments.get(part)
+                                ).exceptionally(e -> {
+                                    LOG.error(
+                                            "Exception on updating assignments for [table={}, partition={}]", e, tableView.name(),
+                                            partId
+                                    );
+
+                                    return null;
+                                });
+                            }
+                        }
+                    }
+
+                    return completedFuture(null);
+                } finally {
+                    busyLock.leaveBusy();
+                }
+            }
+
+            @Override
+            public void onError(Throwable e) {
+                LOG.warn("Unable to process data nodes event", e);
+            }
+        };
+    }
+
+    /**
+     * Listener of replicas configuration changes.
+     *
+     * @param replicasCtx Replicas configuration event context.
+     * @return A future, which will be completed, when event processed by listener.
+     */
+    private CompletableFuture<?> onUpdateReplicas(ConfigurationNotificationEvent<Integer> replicasCtx) {
+        if (!busyLock.enterBusy()) {
+            return completedFuture(new NodeStoppingException());
+        }
+
+        try {
+            if (replicasCtx.oldValue() != null && replicasCtx.oldValue() > 0) {
+                DistributionZoneView zoneCfg = replicasCtx.newValue(DistributionZoneView.class);
+
+                List<TableConfiguration> tblsCfg = new ArrayList<>();
+
+                tablesConfiguration.tables().value().namedListKeys().forEach(tblName -> {
+                    if (tablesConfiguration.tables().get(tblName).zoneId().value().equals(zoneCfg.zoneId())) {
+                        tblsCfg.add(tablesConfiguration.tables().get(tblName));
+                    }
+                });
+
+                CompletableFuture<?>[] futs = new CompletableFuture[tblsCfg.size() * zoneCfg.partitions()];
+
+                int furCur = 0;
+
+                for (TableConfiguration tblCfg : tblsCfg) {
+
+                    LOG.info("Received update for replicas number [table={}, oldNumber={}, newNumber={}]",
+                            tblCfg.name().value(), replicasCtx.oldValue(), replicasCtx.newValue());
+
+                    int partCnt = zoneCfg.partitions();
+
+                    int newReplicas = replicasCtx.newValue();
+
+                    byte[] assignmentsBytes = ((ExtendedTableConfiguration) tblCfg).assignments().value();
+
+                    List<Set<Assignment>> tableAssignments = ByteUtils.fromBytes(assignmentsBytes);
```

Comment:
Nope, it's never possible to have empty assignments, except test based scenarios where we use configuration injection. And it's not worth to fix it because @kgusakov will remove assignments from the configuration within his current ticket https://issues.apache.org/jira/browse/IGNITE-19405